### PR TITLE
PRMP 763 - New cron job lambda to update patient ods weekly

### PR DESCRIPTION
**Pull request by steph-torres-nhs**

### CODE REVIEW

File: infrastructure/lambda-update-patient-ods.tf [hunk 1]
```diff
@@ -0,0 +1,97 @@
+module "update-patient-ods-lambda" {
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:1]: Underscores please :)

File: infrastructure/lambda-update-patient-ods.tf [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    WORKSPACE                  = terraform.workspace
+    LLOYD_GEORGE_DYNAMODB_NAME = "${terraform.workspace}_${var.lloyd_george_dynamodb_table_name}"
+    PDS_FHIR_IS_STUBBED        = local.is_sandbox
+  }
+  is_gateway_integration_needed = false
+  is_invoked_from_gateway       = false
+  memory_size                   = 512
+  lambda_timeout                = 900
+
+  depends_on = [
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:29]: Can go because all are implicit.

File: infrastructure/lambda-update-patient-ods.tf [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    module.lloyd_george_reference_dynamodb_table,
+    aws_iam_policy.ssm_access_policy,
+    module.ndr-app-config
+  ]
+}
+
+resource "aws_iam_policy" "dynamodb_policy_scan_lloyd_george" {
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:36]: Remove `policy_` from name

File: infrastructure/lambda-update-patient-ods.tf [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        ],
+        "Resource" : [
+          module.lloyd_george_reference_dynamodb_table.dynamodb_table_arn,
+        ]
+      }
+    ]
+  })
+}
+
+module "update-patient-ods-alarm" {
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:57]: Underscores please :)

File: infrastructure/lambda-update-patient-ods.tf [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  lambda_function_name = module.update-patient-ods-lambda.function_name
+  lambda_timeout       = module.update-patient-ods-lambda.timeout
+  lambda_name          = "update_patient_ods_handler"
+  namespace            = "AWS/Lambda"
+  alarm_actions        = [module.update-patient-ods-alarm-topic.arn]
+  ok_actions           = [module.update-patient-ods-alarm-topic.arn]
+  depends_on           = [module.update-patient-ods-lambda, module.update-patient-ods-alarm-topic]
+}
+
+module "update-patient-ods-alarm-topic" {
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:68]: Underscores please :)

File: infrastructure/lambda-update-patient-ods.tf [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  source                = "./modules/sns"
+  sns_encryption_key_id = module.sns_encryption_key.id
+  current_account_id    = data.aws_caller_identity.current.account_id
+  topic_name            = "update-patient-ods--topic"
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:72]: `--`

File: infrastructure/schedules.tf [hunk 7]
```diff
@@ -111,4 +111,33 @@ resource "aws_lambda_permission" "statistical_report_schedule_permission" {
...
+  target_id = "update_patient_ods_schedule"
+
+  arn = module.update-patient-ods-lambda.lambda_arn
+  depends_on = [
+    module.update-patient-ods-lambda,
+    aws_cloudwatch_event_rule.update_patient_ods_schedule
+  ]
+}
+
+resource "aws_lambda_permission" "update_patient_ods_schedule_permission" {
```
Comment (chrisbloe-nhse) [inline infrastructure/schedules.tf:133]: Remove `_permission`

File: infrastructure/lambda-update-patient-ods.tf [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+resource "aws_iam_policy" "dynamodb_policy_scan_lloyd_george" {
+  name = "${terraform.workspace}_${var.lloyd_george_dynamodb_table_name}_scan_policy"
+  path = "/"
+
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:43]: Remove extra new line, please :)

File: infrastructure/lambda-update-patient-ods.tf [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      {
+        "Effect" : "Allow",
+        "Action" : [
+          "dynamodb:Scan",
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:41]: Best take the training comma off.

File: infrastructure/lambda-update-patient-ods.tf [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        ],
+        "Resource" : [
+          module.lloyd_george_reference_dynamodb_table.dynamodb_table_arn,
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:44]: Best take the training comma off.

File: infrastructure/lambda-update-patient-ods.tf [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  delivery_policy = jsonencode({
+    "Version" : "2012-10-17",
+    "Statement" : [
+      {
+        "Effect" : "Allow",
+        "Principal" : {
+          "Service" : "cloudwatch.amazonaws.com"
+        },
+        "Action" : [
+          "SNS:Publish",
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-update-patient-ods.tf:78]: Best take the training comma off.

**Review by chrisbloe-nhse — CHANGES_REQUESTED** [review]
Thanks for making the suggested changes. The plan, whilst big, looks ok.

Does this need applying to all envs?

**Review by chrisbloe-nhse — APPROVED** [review]
The commas have been dropped, so this looks good now. Thanks :)

File: infrastructure/lambda-update-patient-ods.tf [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    aws_iam_policy.dynamodb_scan_lloyd_george.arn,
+    aws_iam_policy.ssm_access_policy.arn
+  ]
+  rest_api_id       = null
+  api_execution_arn = null
+
+  lambda_environment_variables = {
+    APPCONFIG_APPLICATION      = module.ndr-app-config.app_config_application_id
+    APPCONFIG_ENVIRONMENT      = module.ndr-app-config.app_config_environment_id
+    APPCONFIG_CONFIGURATION    = module.ndr-app-config.app_config_configuration_profile_id
```
Comment (NogaNHS) [inline infrastructure/lambda-update-patient-ods.tf:19]: Do we need this?

File: infrastructure/schedules.tf [hunk 13]
```diff
@@ -111,4 +111,33 @@ resource "aws_lambda_permission" "statistical_report_schedule_permission" {
     module.statistical-report-lambda,
     aws_cloudwatch_event_rule.statistical_report_schedule
   ]
+}
+
+resource "aws_cloudwatch_event_rule" "update_patient_ods_schedule" {
```
Comment (NogaNHS) [inline infrastructure/schedules.tf:116]: Maybe add a count on this, so it would only run on prod. I think we don't want this to run on dev every week.

**Pull request closed without merging**